pkg/registry/core/service/storage/alloc.go

```diff
@@ -399,7 +402,20 @@ func (al *Allocators) allocIPs(service *api.Service, toAlloc map[api.IPFamily]st
 			allocator = allocator.DryRun()
 		}
 		if ip == "" {
-			allocatedIP, err := allocator.AllocateNext()
+			var allocatedIP net.IP
+			var err error
+			if utilfeature.DefaultFeatureGate.Enabled(features.MultiCIDRServiceAllocator) {
+				// if dryRun is being used the allocator is implemented by ipallocator.dryRunRange
+				// TODO: modify the Interface for the ipallocator
+				svcAllocator, ok := allocator.(*ipallocator.Allocator)
```

Comment:
L400 already does the get allocator by family.  if you modify 
`serviceIPAllocatorsByFamily[...]`
to 
`func getServiceAllocatorByFamily(...) allocator` then
you won't need  `svcAllocator, ok := allocator.(*ipallocator.Allocator)`  nor `the feature gate`.

`func getServiceAllocatorByFamily(...) allocator` can return an allocator that either calls old/new and move all functions to pass in a service ref.

Reply:
I think that is much better, and we should move to that in beta, but at this stage I think that having it this way will be easier to debug and to understand , despite is much verbose ...

Reply:
I misunderstood, I see that way is simpler

Reply:
However, this will need another change when we move to multiple ServiecCIDRs https://github.com/kubernetes/kubernetes/pull/116516/commits/6f60e274e82f5ba5a94f12819ca3cc98fc352cb2 so if you don't mind I will not update it now

Reply:
ack - I will add it to my list for cidrs PR.